thunderbird/processes/wps_update_metadata.py

```diff
@@ -0,0 +1,135 @@
+                supported_formats=[FORMATS.NETCDF],
+            ),
+            LiteralInput(
+                "updates",
+                "Updates File(yaml)",
+                abstract="The filepath of an updates file that specifies what to do to the metadata it finds in the NetCDF file",
+                min_occurs=1,
+                max_occurs=1,
+                data_type="string",
+            ),
+        ]
+        outputs = [
+            ComplexOutput(
+                "output",
+                "Output",
+                abstract="Updated NetCDF files",
+                as_reference=True,
+                supported_formats=[FORMATS.NETCDF],
+            ),
+        ]
+
+        super(UpdateMetadata, self).__init__(
+            self._handler,
+            identifier="update_metadata",
+            title="Update NetCDF Metadata",
+            abstract="Update file containing missing, invalid, or incorrectly named global or variable metadata attributes",
+            store_supported=True,
+            status_supported=True,
+            inputs=inputs,
+            outputs=outputs,
+        )
+
+
+    def copy_and_get_filepath(self, request):
+        '''
+        This function takes an input "request" and returns a filepath to the input data.
+        As the update_metadata simply updates the original file "in place", copying the
+        input data is necessary for two reasons.
+        
+        1. The original file is maintained without any corruption
+        2. Writing back to OPeNDAP file is nearly impossible
+
+        '''
+        if "opendap" in request.inputs: 
+            url = request.inputs["opendap"][0].url
+            input_dataset = xr.open_dataset(url)
+
+            filename = url.split("/")[-1]
+            original = os.path.join(self.workdir, filename)
+            input_dataset.to_netcdf(original, format = "NETCDF4_CLASSIC")
+
+        elif "netcdf" in request.inputs:
+            original = request.inputs["netcdf"][0].file
+
+        else:
+            raise ProcessError(
+                f"You must provide a data source (opendap/netcdf). Inputs provided: {request.inputs}"
+            )
+
+        copy = original[:-3] + "_copy.nc"
+        shutil.copyfile(original, copy)
+        return copy
+        
+
+    def _handler(self, request, response):
+        response.update_status("Starting Process", 0)
+
+        filepath = self.copy_and_get_filepath(request)
+        response.update_status(f"Processing {filepath}", 10)
+
+        updates = request.inputs["updates"][0].data
+
+        try:
+            with open(updates) as ud:
+                updates_instruction = yaml.safe_load(ud)
+
+        except FileNotFoundError:
+            updates_yaml = os.path.join(self.workdir, "updates.yaml")
+    
+            with open(updates_yaml, "w+") as yamlfile:
+                yamlfile.write(updates)
+            with open(updates_yaml) as ud:
+                updates_instruction = yaml.safe_load(ud)
```

Comment:
Would you be able to explain the purpose of this section? Is this meant to handle a string input OR a file input?

Reply:
Additionally, the logic isn't really correct. Both of the file opens inside the exception handling block could *also* raise the same exception. I believe you need to rethink the program flow here.

Reply:
@jameshiebert the `updates_yaml` is a file that has been created inside `tmp` directory as shown in `line 121`, which is inside the `except FileNotFountError` block. The block is to handle when `updates` input is not passed as a path to a file. If no file is found the program creates an empty yaml file and writes the passed input(which is a string) into the file, so the newly created `update.yaml` can be processed in `process_updates`. I did not consider further errors since `update_metadata` would handle them.

Reply:
@sum1lim I think the thing that you may  not have considered here is the fact that if you receive YAML input as a string, you don't have to write it to a file.

`yaml.safe_load()` will happily take a string as input.

```python
>>> yaml.safe_load("""
... - Foo
... - Bar
... - Blah
... """)
['Foo', 'Bar', 'Blah']
```

Your code here takes a string input, writes it to a file, reads from the file to get the original string back. You're doing do extra steps to get back exactly what you started with :) So I'd suggest two thing:

1. Be explicit about what the input is. Is it a file path or is it a string? Simply testing for a FileNotFound error is *implicit* and makes the code more confusing (case in point: it confused both Nik and I as to what you were trying to accomplish).
1. For the case of string YAML input, just use it. Don't bother with the file write/read.